src/openms/include/OpenMS/ANALYSIS/ID/FalseDiscoveryRate.h

```diff
@@ -114,6 +158,488 @@ namespace OpenMS
+    }
+
+    template<class ...Args>
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const PeptideIdentification & id,
+        bool all_hits,
+        Args&& ... args) const
+    {
+      if (all_hits)
+      {
+        for (const PeptideHit &hit : id.getHits())
+        {
+          getScores_(scores_labels, hit, std::forward<Args>(args)...);
+        }
+      }
+      else
+      {
+        //TODO for speed I assume that they are sorted and first = best.
+        //id.sort();
+        const PeptideHit& hit = id.getHits()[0];
+        getScores_(scores_labels, hit, std::forward<Args>(args)...);
+      }
+    }
+
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const PeptideHit & hit,
+        int charge) const
+    {
+        if (charge == hit.getCharge())
+        {
+          checkTDAnnotation_(hit);
+          scores_labels.emplace_back(hit.getScore(), getTDLabel_(hit));
+        }
+    }
+
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const PeptideIdentification & id,
+        int charge) const
+    {
+      for (const PeptideHit &hit : id.getHits())
+      {
+        getScores_(scores_labels, hit, charge);
+      }
+    }
+
+    template<typename HitType, typename std::enable_if<IsHitType<HitType>::value>::type* = nullptr>
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const HitType & hit) const
+    {
+        checkTDAnnotation_(hit);
+        scores_labels.emplace_back(hit.getScore(), getTDLabel_(hit));
+    }
+
+    template<typename IDType, typename std::enable_if<IsIDType<IDType>::value>::type* = nullptr>
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const IDType & id) const
+    {
+      for (const typename IDType::HitType &hit : id.getHits())
+      {
+        getScores_(scores_labels, hit);
+      }
+    }
+
+    template<class ...Args>
+    void getScores_(
+        std::vector<std::pair<double,bool>>& scores_labels,
+        const std::vector<PeptideIdentification> & ids,
+        Args&& ... args) const
+    {
+      for (const auto& id : ids)
+      {
+        getScores_(scores_labels, id, std::forward<Args>(args)...);
+      }
+    }
+
+    void getScores_(std::vector<std::pair<double,bool>>& scores_labels, const std::vector<PeptideIdentification> & targets, const std::vector<PeptideIdentification> & decoys, bool all_hits, int charge, const String& identifier) const
+    {
+
+      for (const PeptideIdentification& id : targets)
+      {
+        if (!identifier.empty() && identifier != id.getIdentifier()) continue;
+
+        if (all_hits)
+        {
+          for (const PeptideHit& hit : id.getHits())
+          {
+            if (charge == 0 || charge == hit.getCharge()) scores_labels.push_back(getScoreLabel_(hit, FalseDiscoveryRate::TrueFunctor<const PeptideHit&>()));
```

Comment:
```suggestion
            if (charge == 0 || charge == hit.getCharge())
            {
              scores_labels.emplace_back(hit.getScore(), FalseDiscoveryRate::TrueFunctor(hit));
            }
```

this is faster and more readable